### PR TITLE
NEW: Add Experimental NGFF Support

### PR DESCRIPTION
**Pull request by John-P**
This PR adds experimental support for NGFF (v0.4). Tested with blosc compressed images generated with https://github.com/John-P/wsic and which have been tested with the https://github.com/ome/ngff info command and napari (https://github.com/ome/napari-ome-zarr). This should also work with NGFF files with blocks compressed using codecs in `imagecodecs.numcodecs`.

Next steps after this is working (future PR):
- Add support for generic Zarr array or a group of resolutions (without NGFF metadata). User can supply an mpp at init.

### CODE REVIEW

File: tiatoolbox/data/remote_samples.yaml [hunk 1]
```diff
@@ -16,6 +16,9 @@ files:
   dicom-1:
     url: [*wsis, "CMU-1.dicom.zip"]
     extract: True
+  ngff-1:
```
Comment (shaneahmed) [inline tiatoolbox/data/remote_samples.yaml:19]: Can we also take a sample from links provided by Glencoe?

File: tiatoolbox/wsicore/metadata/ngff.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
+"""Generation of metadata for the OME-NGFF (zarr) slides.
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/metadata/ngff.py:1]: Do we need to import this in `metadata/__init__.py`?

File: tiatoolbox/wsicore/wsireader.py [hunk 3]
```diff
@@ -2631,6 +2694,175 @@ def read_bounds(
         return utils.transforms.background_composite(image=im_region)
 
 
+class NGFFWSIReader(WSIReader):
+    """Reader for NGFF WSI zarrs.
+
+    Support is currently experimental.
+    """
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:2701]:
```suggestion
    """
    
    
```
```suggestion
    """
```

Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:2701]: new line

**Comment by shaneahmed** [comment]
Please can you address the comments and resolve conflicts. The rest seems fine to me.

File: tiatoolbox/wsicore/wsireader.py [hunk 4]
```diff
@@ -68,6 +69,60 @@ def is_tiled_tiff(path: pathlib.Path) -> bool:
...
+    path = pathlib.Path(path)
+    try:
+        _ = zarr.open(path, mode="r")
+        return True
+    # skipcq: PYL-W0703
+    except Exception:  # noqa: PIE786
+        return False
+
+
+def is_ngff(path: pathlib.Path) -> bool:
```
Comment (ghadjigeorghiou) [inline tiatoolbox/wsicore/wsireader.py:93]: This is not used anywhere. @John-P were you planning on using this when deciding whether to use the NGFFWSIReader or not?

**Review by ghadjigeorghiou — COMMENTED** [review]
Great work John, looks good to me. Just one minor comment to consider. Also, I was wondering if we are planning to support only zarr files that are in the NGFF format? If yes, which version?

File: tiatoolbox/wsicore/wsireader.py (hunk 4 above)
Comment (John-P) [reply tiatoolbox/wsicore/wsireader.py:93]: Yes, that plan is to use this to validate if it is an NGFF formatted Zarr or not.

**Comment by John-P** [comment]
> Great work John, looks good to me. Just one minor comment to consider. Also, I was wondering if we are planning to support only zarr files that are in the NGFF format? If yes, which version?

Version 0.4, see original comment at the top. I will add this to the class docstring.

File: tiatoolbox/data/remote_samples.yaml (hunk 1 above)
Comment (John-P) [reply tiatoolbox/data/remote_samples.yaml:19]: I don't think they had samples for v0.4 last time I checked but I will try to check again

File: tiatoolbox/wsicore/metadata/ngff.py (hunk 2 above)
Comment (John-P) [reply tiatoolbox/wsicore/metadata/ngff.py:1]: I don't expect it to be used outside of the ready function but could do.

File: tiatoolbox/wsicore/wsireader.py (hunk 4 above)
Comment (John-P) [reply tiatoolbox/wsicore/wsireader.py:93]: I may also add support for loading a generic Zarr array also (non NGFF formatted).

File: tiatoolbox/data/remote_samples.yaml (hunk 1 above)
Comment (John-P) [reply tiatoolbox/data/remote_samples.yaml:19]: The latest brightfield sample I have is on v0.1. I believe that Glencoe mentioned that there should be some out there last time we spoke. I'll check emails etc to see if I can find a source of version 0.4 images.

File: tiatoolbox/wsicore/wsireader.py (hunk 4 above)
Comment (John-P) [reply tiatoolbox/wsicore/wsireader.py]: This is now being used to check the Barr in `WSIReader.open`

File: tiatoolbox/data/remote_samples.yaml (hunk 1 above)
Comment (ghadjigeorghiou) [reply tiatoolbox/data/remote_samples.yaml:19]: The latest update on NGFF was in December in this [blog post](https://www.openmicroscopy.org/2021/12/16/ome-ngff.html). @John-P there are some samples that are of version 0.3 at the end of the post.

**Review by ghadjigeorghiou — APPROVED** [review]

File: tiatoolbox/wsicore/wsireader.py [hunk 5]
```diff
@@ -68,6 +70,72 @@ def is_tiled_tiff(path: pathlib.Path) -> bool:
...
+    Returns:
+        bool:
+            True if the file is a Zarr file.
+
+    """
+    path = pathlib.Path(path)
+    try:
+        _ = zarr.open(path, mode="r")
+        return True
+    # skipcq: PYL-W0703
```
Comment (shaneahmed) [inline tiatoolbox/wsicore/wsireader.py:89]:
```suggestion
 
```

**Comment by John-P** [comment]
@shaneahmed I believe this is ready. The coverage issue should be resolved after changes in #386 are merged in.

**Pull request merged**